imaids/utils.py

```diff
@@ -131,6 +136,62 @@ def cosine_function(z, bamp, freq, phase):
     return bamp*_np.cos(freq*z + phase)
 
 
+def get_beff_from_model(model, period, polarization, hmax, x):
+    """Calculate effective field amplitude from model.
+
+    Args:
+        model (radia object): ID model.
+        period (float): ID period.
+        polarization (string): polarization of radiation. (['hp', 'vp', 'cp'])
+        hmax (int): max field harmonic to be considered.
+        x (float): horizontal position to calc field [mm]
+
+    Returns:
+        float : effective field amplitude [T]
+        float: first harmonic field amplitude [T]
+        numpy.ndarray: field along longitudinal positions [T]
+    """
+    zmin = -2*period
+    zmax = 2*period
+    npts = 201
+    z = _np.linspace(zmin, zmax, npts)
+    bvec = model.get_field(x=x,z=z)
+    if polarization == 'vp':
+        b = bvec[:, 0]
+    elif polarization in ('hp', 'cp'):
+        b = bvec[:, 1]
+
+    freq0 = 2*_np.pi/period
+    hs = _np.array(range(1, hmax+1, 2))
+    freqs = hs*freq0
+    amps,*_ = fit_fourier_components(b, freqs, z)
+    beff = _np.sqrt(_np.sum(_np.power(amps/hs, 2)))
+    return beff, amps[0], b
+
```

Comment:
I believe this calculation makes sense for an oscillating field device (right?), so it would probably makes more sense to include this as a method of the SinusoidalFieldSource class (of which cassettes and insertion devices are derived).
At such class, there is a calc_cosine_amplitude method which does partially what this method does. Do you think that if this get_beff_from_model method could be moved to there and use calc_cosine_amplitude internally to avoid a little of code repetition? One further advantage is that at a sinuosidal object the period length is already an attribute.

Reply:
Hi Lucas, turning this function in a method of SinusoidalFieldSource is a very good idea, when we want to get some information about an ID it's indeed more intuitive to call a method of the object ID whereas call a function of another module. I am implementing this change.